File: android/sharedui/components/HoverButton.kt
package com.zhiqihuayun.sharedui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.zhiqihuayun.foundation.design.AppColor
import com.zhiqihuayun.foundation.design.AppFont
import com.zhiqihuayun.foundation.design.AppSpace

/** 悬浮钮视觉 token：与 iOS `HoverButton.Metrics`、design-spec §02 对齐。 */
private object HoverTokens {
    /** 悬浮钮高度基线（dp；icon-only 圆钮直径 = 胶囊高度 = 40）。 */
    val size = 40.dp
    /** 胶囊水平内边距 = AppSpace.lg（16dp）。 */
    val pillInsetX = AppSpace.lg
    /** icon 与 text 间距 = AppSpace.sm（8dp）。 */
    val iconGap = AppSpace.sm
    /** 默认图标字符（U+271A，零图片依赖）。 */
    const val defaultIcon = "✚"
    /** 圆角 = 高度一半（icon-only 圆钮 40→20；胶囊高 40→20，即 full）。 */
    val corner = 20.dp
}

/**
 * HoverButton 悬浮按钮：页面角落常驻的"单功能悬浮动作钮"。
 *
 * 与悬浮族划界：多入口列表面板 → FixedNav；滚动超阈值回顶 → BackTop；
 * 本组件 = 单钮常驻（无滚动阈值自动显隐）、图标字符 + 可选文本、点击交回 onTap。
 *
 * 形态由内容驱动（P3-A）：text 为空 = icon-only 圆钮（40×40，icon 缺省 "✚"）；
 * text 非空 = 胶囊（高 40，宽 = 内容 + 双侧 padding lg），icon 可缺省=纯文本胶囊。
 * 位置由宿主摆放（组件自身 wrap 内容不撑满父级，宿主用 offset/Box 锚定）。
 *
 * @param modifier 宿主修饰（尺寸/位置/点击命中区域）。
 * @param icon 图标字符（单字符/emoji）；null 且 text 为空时兜底默认 "✚"。
 * @param text 文本；null/空 = icon-only 圆钮形态。
 * @param onTap 点击回调（业务自理动作；一期无涟漪/缩放=组件库悬浮族惯例）。
 */
@Composable
fun HoverButton(
    modifier: Modifier = Modifier,
    icon: String? = null,
    text: String? = null,
    onTap: () -> Unit,
) {
    val hasText = !text.isNullOrBlank()
    val effIcon = if (!icon.isNullOrEmpty()) icon else if (hasText) null else HoverTokens.defaultIcon
    val shape = RoundedCornerShape(HoverTokens.corner)
    val inner: Modifier = if (hasText) {
        Modifier.height(HoverTokens.size).padding(horizontal = HoverTokens.pillInsetX)
    } else {
        Modifier.size(HoverTokens.size)
    }
    Box(
        modifier = modifier
            .clip(shape)
            .background(AppColor.primary)
            .clickable(
                interactionSource = remember { MutableInteractionSource() },
                indication = null,
            ) { onTap() },
        contentAlignment = Alignment.Center,
    ) {
        Row(
            modifier = inner,
            verticalAlignment = Alignment.CenterVertically,
            // icon-only：行被强制 40×40，内容需水平居中（Arrangement.Start 会把 ✚ 顶到左侧=未水平居中）；
            // pill：icon+text 中间隔 spacing；纯文本胶囊内容自带 padding 居中即可。
            horizontalArrangement = when {
                hasText && effIcon != null -> Arrangement.spacedBy(HoverTokens.iconGap)
                effIcon != null && !hasText -> Arrangement.Center
                else -> Arrangement.Start
            },
        ) {
            if (effIcon != null) {
                Text(
                    text = effIcon,
                    color = Color.White,
                    fontSize = AppFont.sizeLg,
                    fontWeight = FontWeight.Bold,
                )
            }
            if (hasText && text != null) {
                Text(
                    text = text,
                    color = Color.White,
                    fontSize = AppFont.sizeSm,
                )
            }
        }
    }
}

File: ios/SharedUI/Components/HoverButton.swift
/// HoverButton 悬浮按钮：页面角落常驻的"单功能悬浮动作钮"。
///
/// 与悬浮族划界：多入口列表面板 → FixedNav；滚动超阈值回顶 → BackTop；
/// 本组件 = 单钮常驻（无滚动阈值自动显隐）、图标字符 + 可选文本、点击交回 onTap。
/// 位置由宿主锚定（组件为纯内容视图：self 尺寸 = 钮内容，宿主只锚边/角即可撑开）。
import UIKit

final class HoverButton: UIView {
    /// 设计 token（与 docs/design-spec/hover-button-design-spec.html §02 对齐）。
    private enum Metrics {
        /// 悬浮钮高度基线（pt；icon-only 圆钮直径 = 胶囊高度 = 40）。
        static let size: CGFloat = 40
        /// 胶囊水平内边距 = AppSpace.lg（16pt）。
        static let pillInsetX: CGFloat = AppSpace.lg
        /// icon 与 text 间距 = AppSpace.sm（8pt）。
        static let iconGap: CGFloat = AppSpace.sm
        /// 默认图标字符（U+271A，零图片依赖）。
        static let defaultIcon = "✚"
        /// 按压压暗透明度（iOS 原生按压反馈，差异表放行）。
        static let pressedAlpha: CGFloat = 0.75
    }

    // MARK: - 对外

    let icon: String?
    let text: String?
    /// 点击回调（业务自理动作；nil = 仅形态预览）。
    var onTap: (() -> Void)?

    init(icon: String? = nil, text: String? = nil, onTap: (() -> Void)? = nil) {
        self.icon = icon
        self.text = text
        self.onTap = onTap
        super.init(frame: .zero)
        setup()
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - 内部

    private let bgView = UIView()
    private let content = UIStackView()
    private let iconLabel = UILabel()
    private let textLabel = UILabel()
    private var tapButton: UIButton!
    private var hasText: Bool {
        !(text?.isEmpty ?? true)
    }

    private var effectiveIcon: String? {
        if let icon, !icon.isEmpty { return icon }
        return hasText ? nil : Metrics.defaultIcon
    }

    private func setup() {
        backgroundColor = .clear

        // 主色圆底（圆角按实际 bounds 动态=全库悬浮钮先例）。
        bgView.backgroundColor = AppColor.primary
        bgView.clipsToBounds = true
        bgView.isUserInteractionEnabled = false
        addSubview(bgView)

        // icon + text（仅存在时加入 content），spacing 8pt。
        content.axis = .horizontal
        content.alignment = .center
        content.spacing = Metrics.iconGap
        content.isUserInteractionEnabled = false
        if let effIcon = effectiveIcon {
            iconLabel.text = effIcon
            iconLabel.font = .systemFont(ofSize: AppFont.sizeLg, weight: .bold)
            iconLabel.textColor = .white
            iconLabel.textAlignment = .center
            content.addArrangedSubview(iconLabel)
        }
        if hasText, let text {
            textLabel.text = text
            textLabel.font = .systemFont(ofSize: AppFont.sizeSm)
            textLabel.textColor = .white
            textLabel.textAlignment = .center
            content.addArrangedSubview(textLabel)
        }
        bgView.addSubview(content)

        bgView.snp.makeConstraints { make in
            make.edges.equalToSuperview()
        }
        content.snp.makeConstraints { make in
            make.center.equalToSuperview()
            make.top.greaterThanOrEqualToSuperview()
            make.bottom.lessThanOrEqualToSuperview()
            make.leading.greaterThanOrEqualToSuperview()
            make.trailing.lessThanOrEqualToSuperview()
        }

        // 点击承载（BackTop 同款：常驻不移除，只负责点击与按压反馈）。
        let button = UIButton(type: .custom)
        button.backgroundColor = .clear
        button.addTarget(self, action: #selector(pressDown), for: [.touchDown, .touchDragEnter])
        button.addTarget(self, action: #selector(pressRestore), for: [.touchDragExit, .touchCancel, .touchUpOutside])
        button.addTarget(self, action: #selector(didTap), for: .touchUpInside)
        addSubview(button)
        tapButton = button
        button.snp.makeConstraints { make in
            make.edges.equalToSuperview()
        }
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        // 圆角 = bounds 高一半（icon-only 圆钮 40×40 → 20；胶囊高 40 → 20，即 full）。
        bgView.layer.cornerRadius = bounds.height / 2
    }

    override var intrinsicContentSize: CGSize {
        let base: CGFloat = Metrics.size
        if !hasText {
            return CGSize(width: base, height: base)
        }
        // 胶囊宽 = 内容自然宽（icon + gap + text）+ 双侧 padding lg。
        var contentWidth: CGFloat = 0
        if effectiveIcon != nil { contentWidth += iconLabel.intrinsicContentSize.width }
        if hasText { contentWidth += textLabel.intrinsicContentSize.width }
        if effectiveIcon != nil, hasText { contentWidth += Metrics.iconGap }
        return CGSize(width: contentWidth + Metrics.pillInsetX * 2, height: base)
    }

    @objc private func pressDown() {
        UIView.animate(withDuration: 0.12) { [weak self] in
            self?.bgView.alpha = Metrics.pressedAlpha
        }
    }

    @objc private func pressRestore() {
        UIView.animate(withDuration: 0.12) { [weak self] in
            self?.bgView.alpha = 1
        }
    }

    @objc private func didTap() {
        UIView.animate(withDuration: 0.12) { [weak self] in
            self?.bgView.alpha = 1
        }
        onTap?()
    }
}
